pwiz_tools/Skyline/Controls/Graphs/GraphSpectrum.cs

```diff
@@ -911,6 +987,240 @@ private void ClearGraphPane()
+
+                                var replicateName = matchingFile.Chromatograms.Name;
+                                spectraRedundant.Add(new SpectrumDisplayInfo(spectrumInfo, precursor.Precursor,
+                                    replicateName, matchingFile.FilePath, matchingFile.FileOrder,
+                                    spectrumInfo.RetentionTime, false));
+
+                                // Include the best spectrum twice, once displayed in the normal
+                                // way and once displayed with its replicate and retention time.
+                                if (spectrumInfo.IsBest)
+                                {
+                                    var iBest = precursorSpectra[i].Spectra.IndexOf(s =>
+                                        Equals(s.Name, spectrumInfo.Name) &&
+                                        Equals(s.LabelType, spectrumInfo.LabelType));
+                                    if (iBest != -1)
+                                    {
+                                        precursorSpectra[i].Spectra[iBest] = new SpectrumDisplayInfo(
+                                            precursorSpectra[i].Spectra[iBest].SpectrumInfo, precursor.Precursor,
+                                            replicateName,
+                                            matchingFile.FilePath, 0, spectrumInfo.RetentionTime, true);
+                                    }
+                                }
+
+                                if (!dictReplicateNameFiles.TryGetValue(replicateName, out var setFiles))
+                                {
+                                    setFiles = new HashSet<string>();
+                                    dictReplicateNameFiles.Add(replicateName, setFiles);
+                                }
+
+                                setFiles.Add(spectrumInfo.FilePath);
+                            }
+
+                            // Determine if replicate name is sufficient to uniquely identify the file
+                            foreach (var spectrumInfo in spectraRedundant)
+                            {
+                                var replicateName = spectrumInfo.ReplicateName;
+                                if (replicateName != null && dictReplicateNameFiles[replicateName].Count < 2)
+                                    spectrumInfo.IsReplicateUnique = true;
+                            }
+                            spectraRedundant.Sort();
+                            precursorSpectra[i].Spectra.AddRange(spectraRedundant);
+                        }
+                    }
+                    catch (Exception)
+                    {
+                        // ignored
+                    }
+                }
+
+                _spectraLoadCache = Cache.Create(null, precursorSpectra, settings);
+            }
+
+            public bool GetPrecursorStrings(IReadOnlyList<PrecursorSpectra> spectra, out string[] outStrings)
```

Comment:
I would recommend using "ImmutableList&lt;string>" instead of "string[]" since it would prevent the elements in the list from being changed.